mass_mailing_partner/models/mailing_contact.py

```diff
@@ -34,47 +34,75 @@ def _check_partner_id_list_ids(self):
                     )
 
     @api.onchange("partner_id")
-    def _onchange_partner_mass_mailing_partner(self):
+    def _onchange_partner_mass_mailing_partner(self, change_tag_bool=True):
         if self.partner_id:
             self.name = self.partner_id.name
             self.email = self.partner_id.email
             self.title_id = self.partner_id.title
-            self.company_name = self.partner_id.company_id.name
+            self.company_name = (
+                self.partner_id.company_id.name or self.partner_id.company_name
+            )
             self.country_id = self.partner_id.country_id
             category_ids = self.partner_id.category_id
-            if category_ids:
+            if category_ids and change_tag_bool:
                 self.tag_ids = category_ids
 
-    @api.model
-    def create(self, vals):
-        record = self.new(vals)
-        if not record.partner_id:
-            record._set_partner()
-        record._onchange_partner_mass_mailing_partner()
-        new_vals = record._convert_to_write(record._cache)
-        new_vals.update(
-            subscription_list_ids=vals.get("subscription_list_ids", []),
-            list_ids=vals.get("list_ids", []),
-        )
-        return super().create(new_vals)
+    @api.model_create_multi
+    def create(self, vals_list):
+        new_vals_list = []
+        for vals in vals_list:
+            record = self.new(vals)
+            if not record.partner_id:
+                record._set_partner()
+            record._onchange_partner_mass_mailing_partner(change_tag_bool=False)
+            new_vals = record._convert_to_write(record._cache)
+            new_vals.update(
+                subscription_list_ids=vals.get("subscription_list_ids", []),
+                list_ids=vals.get("list_ids", []),
+            )
+            # Don't take the tag_ids from the cache, because it triggers to re-update
+            # the values of creating new Partner by removing the key `category_id`.
+            # Therefore, new partner does not have value for Tags
+            tag_vals = vals.get("tag_ids")
+            if record.partner_id.category_id:
+                # Get Tags from Partner to take into account the tags from Mailing List as well
+                tag_vals = [(6, 0, record.partner_id.category_id.ids)]
+            new_vals.update({"tag_ids": tag_vals})
```

Comment:
Is this really necessary?
Can you add a test to verify it?
Can you homogenize this code and the write code in a new method?

Reply:
Of course it is necessary, i don't add these lines for fun. You can play with this by removing them, then compare the result.
the test has been done at this line: https://github.com/OCA/social/blob/7644f8f70daf10f4ea3f9708af973660b27b106e/mass_mailing_partner/tests/base.py#L66

>  Can you homogenize this code and the write code in a new method?

Good idea, WIP on it

Reply:
@victoralmau:  Updated